### PR TITLE
Translations from the config XML files 

### PR DESCRIPTION
**Pull request by ifarzana**
- Related to https://github.com/xibosignageltd/xibo-private/issues/183

### CODE REVIEW

**Comment by ifarzana** [comment]
@dasgarner I am getting an error `PHP Fatal error:  Uncaught LogicException: Unable to register extension "Slim\Views\TwigExtension" as it is already registered. in /var/www/cms/vendor/twig/twig/src/ExtensionSet.php:132
Stack trace:` 

Do you see anything that I am not doing correctly?

**Comment by dasgarner** [comment]
I don't think creating the whole application stack is the right way to pursue this. We discussed creating the ModuleFactory and ModuleTemplateFactory manually instead?

To do that you'd need to use `new ModuleFactory()` and supply mocked objects for its dependencies.

I think this would be workable because the bit we want (the XML parsing) doesn't rely on any of those dependencies (I think)

File: bin/locale.php [hunk 1]
```diff
@@ -1,4 +1,5 @@
 <?php
+
 /**
  * Copyright (C) 2019 Xibo Signage Ltd
```
Comment (dasgarner) [inline bin/locale.php:4]: Odd that this copyright didn't update? is the profile setup in your IDE?

File: bin/locale.php [hunk 2]
```diff
@@ -38,6 +40,10 @@
 define('PROJECT_ROOT', realpath(__DIR__ . '/..'));
 require_once PROJECT_ROOT . '/vendor/autoload.php';
 
+if (!file_exists(PROJECT_ROOT . '/web/settings.php')) {
```
Comment (dasgarner) [inline bin/locale.php:43]: I don't think we need this?

File: bin/locale.php [hunk 3]
```diff
@@ -55,14 +61,155 @@
...
+// Mock Config Service
+class MockConfigService extends ConfigService
+{
+    public function getSetting($setting, $default = null, $full = false)
+    {
+        return '';
+    }
+}
+
+// Transactor funtion
```
Comment (dasgarner) [inline bin/locale.php:90]:
```suggestion
// Translator function
```

File: bin/locale.php [hunk 4]
```diff
@@ -55,14 +61,155 @@
...
+);
+$moduleTemplateFactory->useBaseDependenciesService($baseDepenencyService);
+// Get all module templates
+$moduleTemplates = $moduleTemplateFactory->getAll();
+
+// --------------
+// Create translation file
+// Each line contains title or description or properties of the module/templates
+$file = PROJECT_ROOT. '/locale/moduletranslate.php';
+if (!file_exists($file)) {
```
Comment (dasgarner) [inline bin/locale.php:130]: I don't think you need to do this - file_put_contents will create a file if it doesn't exist and overwrite one if it does.

File: bin/locale.php [hunk 5]
```diff
@@ -55,14 +61,155 @@
...
+                foreach ($setting->options as $option) {
+                    if (!empty($option->title)) {
+                        $content .= "echo __('$option->title');" . PHP_EOL;
+                    }
+                }
+            }
+        }
+
+        // Properties translation
+        foreach ($module->properties as $property) {
```
Comment (dasgarner) [inline bin/locale.php:164]: Sorry, I did add another requirement regarding parsing a rule message :smile:

**Review by dasgarner — CHANGES_REQUESTED** [review]
Looks like you've cracked it - just a few suggestions.

**Review by dasgarner — APPROVED** [review]
Looks good!

**Pull request merged**